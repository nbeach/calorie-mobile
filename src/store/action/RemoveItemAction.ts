import {createActionFactory} from "./action-util"

export interface RemoveItemAction {
    readonly id: string
}

export const removeItemAction = createActionFactory<RemoveItemAction>("REMOVE_ITEM_ACTION")

