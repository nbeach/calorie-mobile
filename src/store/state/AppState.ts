import {ItemState} from "./ItemState"
import {AddItemState} from "./AddItemState"

export interface AppState {
    readonly items: ReadonlyArray<ItemState>,
    readonly addItem: AddItemState
}

