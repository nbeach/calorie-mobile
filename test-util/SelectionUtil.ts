import {ShallowWrapper} from "enzyme"
import {ReactTestRenderer} from "react-test-renderer";

export const nodes = (testId: string, wrapper: ShallowWrapper) => wrapper.findWhere(node => node.prop("testID") === testId)
export const node = (testId: string, wrapper: ShallowWrapper) => nodes(testId, wrapper).first()

export const instances = (testId: string, tree: ReactTestRenderer) => tree.root.findAll(thing => thing.props.testID === testId)
export const instance = (testId: string, tree: ReactTestRenderer) => tree.root.find(thing => thing.props.testID === testId)
