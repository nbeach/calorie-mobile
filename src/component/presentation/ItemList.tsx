import React from "react"
import {StyleSheet, Text, View} from "react-native"
import {ItemState} from "../../store/state/ItemState"
import {border} from "../../util/StyleUtil"

export interface ItemListProps {
    readonly items: ReadonlyArray<ItemState>
}

export const ItemList = (props: ItemListProps) => {
    return <View style={styles.container}>
        {props.items.map(item => <Text key={item.name} style={styles.item}>{item.name}</Text>)}
    </View>
}


const styles = StyleSheet.create({
    container: {
        flex: 1,
        alignItems: "center",
    },
    item: {
        padding: 10,
        width: "100%",
        textAlign: "center",
        fontSize: 18,
        ...border(1, "solid", "#AAAAAA"),
    },

})
