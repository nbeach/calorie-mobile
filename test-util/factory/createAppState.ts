import {AppState} from "../../src/store/state/AppState"
import {createAddItemState} from "./createAddItemState"

export const createAppState = (partial: Partial<AppState> = {}): AppState => {
    const defaults: AppState = {
        items: [],
        totalCalories: 0,
        addItem: createAddItemState(),
    }

    return {...defaults, ...partial}
}
