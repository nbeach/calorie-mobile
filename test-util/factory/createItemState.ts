import {ItemState} from "../../src/store/state/ItemState"

export const createItemState = (partial: Partial<ItemState> = {}): ItemState => {
    const defaults: ItemState = {
        id: "00000000-0000-0000-0000-000000000000",
        name: "",
        calories: -1,
    }

    return {...defaults, ...partial}
}
