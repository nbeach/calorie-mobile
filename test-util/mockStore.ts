import {AppState} from "../src/store/state/AppState"

export const mockStore = (state: AppState) => ({
    getState: () => state,
    dispatch: jest.fn(),
    subscribe: jest.fn(),
    replaceReducer: jest.fn(),
})
