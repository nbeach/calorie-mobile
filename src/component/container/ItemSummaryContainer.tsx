import {ItemListDispatchProps} from "../presentation/ItemList"
import {connect, MapDispatchToPropsFunction, MapStateToProps} from "react-redux"
import {AppState} from "../../store/state/AppState"
import {removeItemAction} from "../../store/action/RemoveItemAction"
import {ItemSummary, ItemSummaryStateProps} from "../presentation/ItemSummary"

export const mapStateToProps: MapStateToProps<ItemSummaryStateProps, void, AppState> = ({totalCalories, items}) => ({ items, totalCalories })

export const mapDispatchToProps: MapDispatchToPropsFunction<ItemListDispatchProps, void> = (dispatch) => ({
    removeItem: (id: string) => dispatch(removeItemAction({ id })),
})

export const ItemSummaryContainer = connect(mapStateToProps, mapDispatchToProps)(ItemSummary)
