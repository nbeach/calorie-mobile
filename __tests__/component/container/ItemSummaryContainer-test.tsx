import React from "react"

import renderer, {ReactTestRenderer} from "react-test-renderer"
import {Provider} from "react-redux"
import {AppState} from "../../../src/store/state/AppState"
import {Store} from "redux"
import {instances} from "../../../test-util/SelectionUtil"
import {ItemListTestIds} from "../../../src/component/presentation/ItemList"
import {removeItemAction} from "../../../src/store/action/RemoveItemAction"
import {createAppState} from "../../../test-util/factory/createAppState"
import {ItemSummaryContainer} from "../../../src/component/container/ItemSummaryContainer";
import {createItemState} from "../../../test-util/factory/createItemState";
import {mockStore} from "../../../test-util/mockStore";

describe(ItemSummaryContainer.name, () => {
    let tree: ReactTestRenderer, store: Store<AppState>

    beforeEach(() => {

        store = mockStore(createAppState({
            items: [
                createItemState({id: "00000000-0000-0000-0000-000000000000", name: "Chicken", calories: 100}),
                createItemState({id: "11111111-1111-1111-1111-111111111111", name: "Beef", calories: 200}),
            ],
            totalCalories: 500,
        }))

        tree = renderer.create(
            <Provider store={store}>
                <ItemSummaryContainer state/>
            </Provider>,
        )
    })

    it("presentation has not changed", () => {
        expect(tree.toJSON()).toMatchSnapshot()
    })

    it("when the remove button is pressed removes the item", () => {
        instances(ItemListTestIds.RemoveButton, tree)[0].props.onPress()
        expect(store.dispatch).toHaveBeenCalledWith(removeItemAction({ id: "00000000-0000-0000-0000-000000000000"}))
    })

})

