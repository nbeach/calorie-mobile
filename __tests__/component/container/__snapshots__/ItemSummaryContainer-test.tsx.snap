// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connect presentation has not changed 1`] = `
<View>
  <View>
    <Text>
      500
    </Text>
  </View>
  <View>
    <View
      style={
        Object {
          "borderBottomWidth": 1,
          "borderColor": "#AAAAAA",
          "borderStyle": "solid",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 18,
          }
        }
      >
        Chicken
      </Text>
      <Text
        style={
          Object {
            "fontSize": 18,
          }
        }
      >
        100
      </Text>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="RemoveButton"
      >
        <Text>
          X
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "borderBottomWidth": 1,
          "borderColor": "#AAAAAA",
          "borderStyle": "solid",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 18,
          }
        }
      >
        Beef
      </Text>
      <Text
        style={
          Object {
            "fontSize": 18,
          }
        }
      >
        200
      </Text>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="RemoveButton"
      >
        <Text>
          X
        </Text>
      </View>
    </View>
  </View>
</View>
`;
